### PR TITLE
Fixes #8 feature request: Add Devise Gem with Configurable

### LINKED ISSUE #8: Add Devise Gem with configurable
Add Devise Gem with configurable. Making sure that it is api's only. You can test the APIs using Postman.

You should pick this task only if you know Postman and know how to create APIs in rails. If you don't know that please watch below videos before assigning yourself.

Reference: https://www.youtube.com/watch?v=Wq9i77BUBlQ
Reference: https://youtu.be/OmT_uNtT7vY

### PR DESCRIPTION
**Pull request by vanamraghu**
1. Updated the code base to have only rails api 
2. Only Devise gem is added with below customization
3. Two controllers registrations & sessions controller is added along with enabling the confirmable (This would send or prompt for confirming the mail address)
4. One bug thats existing in devise for sending fake session to the request headers is provided under concerns (rack_session_fix.rb)
5. cors.rb is uncommented to ensur cross origin is enabled so that external application is provided access
6. In user.rb model, enabled the confirmable 
7. In routes.rb overridden the paths of sessions and registrations
8. Created migration file to enable confirmable
9. Uncommented in schema.rb fields related to confirmable
10. Added in gem file, jsonapi-serializer this is needed to convert or serialize ruby object to JSON
11. User Serializer is added and only id, email and created_at is provided as part of user object
12. Registrations and sessions controller is updated with response_to and response_with so that json is returned (as its rails only api, we are not making use of views and only json response is returned with right status code and message)
13. JSON response messages are modified to have below structure
### Response

> { code: 200, message: 'Signed up successfully', data: { id: 5, email: 'xyz@mail.com', created_at: '2024-08-24T08:00:00.7898"}

Resolves #8

### Description
Summary: Addition & configuration of devise gem
Dependencies: Devise, jsonapi-serializer

### Type of change

- [ ] Feature

### How to test this PR?
Once configuration is completed, below api's can be verified

1. Signup, use below api for signingup/registration, below payload shall be used for sending request and response is displayed below

CleanShot 2024-08-24 at 10 53 13

2. Run the mailcatcher so that configured mail address in devise.rb should receive confirmation mail as shown below

**- Run mailcatcher using this command on terminal:** `mailcatcher`

It should run like below:
```
Starting MailCatcher v0.10.0
==> smtp://127.0.0.1:1025
==> http://127.0.0.1:1080
```

CleanShot 2024-08-23 at 23 04 40

3. Confirm the mail by clicking on the link

CleanShot 2024-08-23 at 23 05 56

4. Once link is clicked, users account would be marked as confirmed and we are good to login 
5. Use below api for login/signin, (expected response should be successful login as provided in the sessions controller)

CleanShot 2024-08-24 at 10 54 17

6. Verify for registration by providing same email id, then we would get error as shown below 

CleanShot 2024-08-24 at 10 54 43

### CODE REVIEW

File: app/controllers/home_controller.rb [hunk 1]
```diff
@@ -0,0 +1,4 @@
+class HomeController < ApplicationController
+  def index; end
+end
+
```
Comment (rt4914) [inline app/controllers/home_controller.rb:4]: Remove this extra blank space.

Also why do we need this controller?
Our backend will have apis only so we don;t need this UI part in rails.

**Review by rt4914 — CHANGES_REQUESTED** [review]
@vanamraghu There was some setup work pending before we can actually work on this PR.

Basically the project needs to be in api only mode.

I have created the PR you can approve, merge and delete the branch and then update your code accordingly.

Basically there should be no UI work in rails.
https://github.com/rt4914/KlinicCon-Backend/pull/29

File: app/controllers/home_controller.rb (hunk 1 above)
Comment (vanamraghu) [reply app/controllers/home_controller.rb]: Completed and removed devise views too, only rails api

**Comment by vanamraghu** [comment]
> @vanamraghu There was some setup work pending before we can actually work on this PR.
> 
> Basically the project needs to be in api only mode.
> 
> I have created the PR you can approve, merge and delete the branch and then update your code accordingly.
> 
> Basically there should be no UI work in rails. #29

@rt4914 
Yes changes are done, branch has only api changes not UI

File: app/models/user.rb [hunk 2]
```diff
@@ -0,0 +1,6 @@
+class User < ApplicationRecord
+  # Include default devise modules. Others available are:
+  # :confirmable, :lockable, :timeoutable, :trackable and :omniauthable
+  devise :database_authenticatable, :registerable,
+         :recoverable, :rememberable, :validatable
```
Comment (rt4914) [inline app/models/user.rb:5]: We need to add `confirmable` too which will also affect the database schema.

File: config/locales/devise.en.yml [hunk 3]
```diff
@@ -0,0 +1,65 @@
+# Additional translations at https://github.com/heartcombo/devise/wiki/I18n
```
Comment (rt4914) [inline config/locales/devise.en.yml:1]: For now we can keep this file but not sure if we will actually use it.

File: db/migrate/20240822013600_devise_create_users.rb [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+      t.datetime :remember_created_at
+
+      ## Trackable
+      # t.integer  :sign_in_count, default: 0, null: false
+      # t.datetime :current_sign_in_at
+      # t.datetime :last_sign_in_at
+      # t.string   :current_sign_in_ip
+      # t.string   :last_sign_in_ip
+
+      ## Confirmable
```
Comment (rt4914) [inline db/migrate/20240822013600_devise_create_users.rb:24]: We need confirmable too.

**Review by rt4914 — CHANGES_REQUESTED** [review]
You will need to add two devise controllers too something like `registrations_controller` and `sessions_controller` and maybe some more controllers for `confirmation` related task.

Also if you see the original video that I attached https://www.youtube.com/watch?v=Wq9i77BUBlQ

Until and unless we don't make the API call from postman we cannot sat that we have configured the Devise gem for APIs correctly.


Also update your PR description.

File: app/models/user.rb (hunk 2 above)
Comment (vanamraghu) [reply app/models/user.rb:5]: Added confirmable in the model

File: db/migrate/20240822013600_devise_create_users.rb (hunk 4 above)
Comment (vanamraghu) [reply db/migrate/20240822013600_devise_create_users.rb:24]: Added as part of a separate migration

**Comment by vanamraghu** [comment]
> You will need to add two devise controllers too something like `registrations_controller` and `sessions_controller` and maybe some more controllers for `confirmation` related task.
> 
> Also if you see the original video that I attached https://www.youtube.com/watch?v=Wq9i77BUBlQ
> 
> Until and unless we don't make the API call from postman we cannot sat that we have configured the Devise gem for APIs correctly.
> 
> Also update your PR description.

@rt4914 Updated the PR description and also about testing

File: app/controllers/users/sessions_controller.rb [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+      status: {code: 200, message: 'Logged in successfully.' },
+      data: UserSerializer.new(resource).serializable_hash[:data][:attributes]
+    }, status: :ok
+  end
+
+  def respond_to_on_destroy
+    if current_user
+      render json: {
+        status: 200,
+        message: "logged out successfully"
```
Comment (rt4914) [inline app/controllers/users/sessions_controller.rb:20]: Write this as `Logged out successfully`

basically first character capital

File: Gemfile [hunk 6]
```diff
@@ -49,3 +49,8 @@ group :development do
   # A gem that calculates the exchange rate using published rates from open-exchange-rates
   gem 'money-open-exchange-rates'
 end
+
+gem 'rack-cors'
+
+gem 'devise', '~> 4.9'
+gem 'jsonapi-serializer'
```
Comment (rt4914) [inline Gemfile:56]: EOF (End of File) is missing, add one blank line at the end.

**Review by rt4914 — CHANGES_REQUESTED** [review]
Really great work @vanamraghu 

One important change:
Current json sturcture is 
```
{
status {
  code ...
  message ...
},
data {
}
```


This should be changed to 

```
{
  code ...
  message ...
  data {
   ....
  }
}
```

We don't need status and also we should keep data as sibling to code and message.

File: app/controllers/users/sessions_controller.rb (hunk 5 above)
Comment (vanamraghu) [reply app/controllers/users/sessions_controller.rb:20]: Changed to Logged

File: Gemfile (hunk 6 above)
Comment (vanamraghu) [reply Gemfile:56]: Fixed

**Comment by vanamraghu** [comment]
> Really great work @vanamraghu
> 
> One important change: Current json sturcture is
> 
> ```
> {
> status {
>   code ...
>   message ...
> },
> data {
> }
> ```
> 
> This should be changed to
> 
> ```
> {
>   code ...
>   message ...
>   data {
>    ....
>   }
> }
> ```
> 
> We don't need status and also we should keep data as sibling to code and message.

@rt4914 
Fixed as per the new changes in json structure

**Review by rt4914 — APPROVED** [review]
LGTM, Thanks.

**Pull request merged**